### DISCUSSION TITLE
springboot with cluster of brokers

### DISCUSSION THREAD

**Discussion by lnthai2002** [arkmq-org/arkmq-org-broker-operator#581] [Q&A] [answered]
Hello,
I am running a single broker but now i am trying to scale up to a cluster of 2 brokers with the following config:

apiVersion: broker.amq.io/v1beta1
kind: ActiveMQArtemis
metadata:
  name: activemq-artemis-broker
spec:
  acceptors:
    - name: "scaleDown"
      bindToAllInterfaces: true
      port: 61616
      protocols: "CORE"
  addressSettings:
    applyRule: merge_all
    addressSetting:
    - match: '#'
      deadLetterAddress: DLQ
      expiryAddress: ExpiryQueue
      redistributionDelay: 60000
      redeliveryDelay: 0
      maxSizeBytes: "-1"
      messageCounterHistoryDayLimit: 10
      addressFullPolicy: PAGE
      autoCreateQueues: true
      autoCreateAddresses: true
      autoCreateJmsQueues: true
      autoCreateJmsTopics: true
      autoDeleteQueues: false
      autoDeleteCreatedQueues: false
      autoDeleteAddresses: false
      defaultConsumerWindowSize: 0
  console:
    expose: true
    sslEnabled: false
  deploymentPlan:
    size: 2
    requireLogin: true
    messageMigration: true
    persistenceEnabled: true
    podSecurityContext:
      fsGroup: 0
    storage:
      size: {{ .Values.persistence.size }}
      storageClassName: {{ .Values.persistence.storageClass }}
    resources:
      requests:
        cpu: {{ .Values.resources.requests.cpu }}
        memory: {{ .Values.resources.requests.memory }}
  adminUser: {{ .Values.artemisUser }}
  adminPassword: {{ .Values.artemisPassword }}

I can see 2 broker pods but only 1 services to port 61616:
activemq-artemis-broker-hdls-svc       ClusterIP   None            <none>        8161/TCP,61616/TCP   13d

My application is a spring boot 3 app with spring-boot-starter-artemis dependency so my application.properties has this when i am running a single broker:
spring.artemis.broker-url=tcp://activemq-artemis-broker-hdls-svc.mynamespace.svc.cluster.local:61616?clientFailureCheckPeriod=30000
However, now that i have a cluster of 2 brokers, do i need to add a 2nd service to the broker-url string so that when a broker goes down, the client can connect to the other broker (i enabled messageMigration above)?

Also, is there a way to change the name of the kubernetes services from activemq-artemis-broker-hdls-svc  to something like "artemis-core-svc"? Asking because I have another standalone broker (not deployed by operator) running and it already has that service name so when i switch to operator i want to keep all the client unchanged.

**Comment by brusdev** [comment]
The operator automatically creates a service for each broker acceptors and you could use those services to build the url of your spring boot app.
In your case the operator is not creating the acceptor services because your accewptor name includes an uppercase character:
```
ERROR package k8s_actions Failed to create new *v1.Service {"ActiveMQArtemis Name": "artemis-broker-scaleDown-0-svc", "Namespace": "default", "error": "Service \"artemis-broker-scaleDown-0-svc\" is invalid: [metadata.name: Invalid value: \"artemis-broker-scaleDown-0-svc\": a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character (e.g. 'my-name',  or 'abc-123', regex used for validation is '[a-z]([-a-z0-9]*[a-z0-9])?'), spec.ports[0].name: Invalid value: \"scaleDown-0\": a lowercase RFC 1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')]"}
```

Changing the acceptor name will fix this issue but I would recommend to use another acceptor port because 61616 is used also for internal purposes. 

The operator creates acceptor name using the following pattern the `<CR_NAME>-<ACCEPTOR_NAME>-<SS_INDEX>-svc` you could create another service if you need a custom name or create an issue to request this new feature.

**Comment by lnthai2002** [comment]
Thank you. I removed the acceptors block from the config so that it is similar to the example https://artemiscloud.io/docs/tutorials/scaleup_and_scaledown/ and  when i looked at the broker.xml in the broker pod i see: 
`<acceptor name="scaleDown">tcp://activemq-artemis-broker-ss-0.activemq-artemis-broker-hdls-svc.ncp-stack-testing12.svc.cluster.local:61616?protocols=CORE;tcpSendBufferSize=1048576;tcpReceiveBufferSize=1048576;useEpoll=true;amqpCredits=1000;amqpMinCredits=300</acceptor>`
so I assume this is the default connector.
However, i still see only 1 kubernetes service that expose port 61616 but 2 services for web console:
`root@rdadmin-Virtual-Machine:/home/rdadmin# kubectl get service -n ncp-stack-testing12
NAME                                   TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                      AGE
activemq-artemis-broker-hdls-svc       ClusterIP   None             <none>        8161/TCP,61616/TCP           21m
activemq-artemis-broker-ping-svc       ClusterIP   None             <none>        8888/TCP                     21m
activemq-artemis-broker-wconsj-0-svc   ClusterIP   172.20.101.81    <none>        8161/TCP,8162/TCP            21m
activemq-artemis-broker-wconsj-1-svc   ClusterIP   172.20.42.7      <none>        8161/TCP,8162/TCP            4m36s`

My question is more about how to config springboot client to use clusted of broker, do i need to specify all services for all broker or just the single service `activemq-artemis-broker-hdls-svc` that look like a load balancer for all the brokers in the cluster?

**Reply by brusdev** [reply, marked as answer]
The operator automatically creates a service for each broker acceptors and you could use those services to build the url of your spring boot app, i.e. the operator will create 2 services for the core-acceptor, one for each broker instance defined in the following CR:

```
apiVersion: broker.amq.io/v1beta1
kind: ActiveMQArtemis
metadata:
  name: artemis-broker
spec:
  deploymentPlan:
    size: 2
  acceptors:
    - name: "core-acceptor"
      port: 61617
      protocols: "CORE"
```

```
$ kubectl get service
NAME                                 TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)              AGE
artemis-broker-core-acceptor-0-svc   ClusterIP   10.99.251.174   <none>        61617/TCP            3m
artemis-broker-core-acceptor-1-svc   ClusterIP   10.111.165.7    <none>        61617/TCP            2m34s
```

The spring boot app could use the following broker url to distribuite client across 2 broker instances and get reconnect after a temporary issue:
```
spring.artemis.broker-url=(tcp://artemis-broker-core-acceptor-0-svc.mynamespace.svc.cluster.local:61617,tcp://artemis-broker-core-acceptor-1-svc.mynamespace.svc.cluster.local:61617)?reconnectAttempts=-1
```

**Comment by lnthai2002** [comment]
thank you

**Comment by lnthai2002** [comment]
Hello, is there a way to scale up the number of broker without changing the connection string on the client side?

**Reply by brusdev** [reply]
@lnthai2002 you could manually create a ClusterIP service, i.e.:
```
apiVersion: broker.amq.io/v1beta1
kind: ActiveMQArtemis
metadata:
  name: artemis-broker
spec:
  deploymentPlan:
    size: 2
  acceptors:
    - name: "core-acceptor"
      port: 61617
      protocols: "CORE"
```

```
oc create -f - <<EOF
apiVersion: v1
kind: Service
metadata:
  name: artemis-broker-svc
spec:
  ports:
  - port: 61617
    protocol: TCP
  selector:
    ActiveMQArtemis: artemis-broker
    application: artemis-broker-app
  type: ClusterIP
status:
  loadBalancer: {}
EOF
```

```
spring.artemis.broker-url=tcp://artemis-broker-svc.mynamespace.svc.cluster.local:61617?reconnectAttempts=-1
```
